### PR TITLE
Anos/feat/approve shop

### PR DESCRIPTION
**Pull request by Anon-136**
Screen Shot 2565-03-28 at 00 00 25

Screen Shot 2565-03-28 at 00 01 11

### CODE REVIEW

File: app/api/uploadURL.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { BlitzApiHandler } from 'blitz'
+
+import {
+  gcpStorageBucketName,
+  gcpStorageClientEmail,
+  gcpStoragePrivateKey,
+  gcpStorageProjectId,
+} from 'app/core/environment'
+
+const handler: BlitzApiHandler = async (req, res) => {
```
Comment (paphonb) [inline app/api/uploadURL.ts:11]: This can be converted into a mutation, so you won't have to handle types yourself.

File: app/api/uploadURL.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  const gc = new Storage({
```
Comment (paphonb) [inline app/api/uploadURL.ts:12]: Might be a good idea to move this outside, and only init the storage once.

File: app/pages/menu.tsx [hunk 3]
```diff
@@ -60,16 +60,31 @@ const Menu: BlitzPage<MenuProps> = ({ user }) => {
...
+              <Link href={Routes.RegisterPage().pathname} passHref>
+                <MenuListItem
+                  as="a"
+                  icon={<FiLayout />}
+                  title="Register new shop"
+                />
+              </Link>
+            </>
+          ) : shop.shopStatus === ShopStatus.REQUESTED ? (
+            <MenuListItem as="a" title="Your registration is in progress" />
```
Comment (paphonb) [inline app/pages/menu.tsx:86]: Maybe switch to div as it's not clickable

```suggestion
            <MenuListItem as="div" title="Your registration is in progress" />
```

File: app/shop/components/UploadImagesBlock.tsx [hunk 4]
```diff
@@ -2,19 +2,53 @@ import { Image, Link } from 'blitz'
...
 export const UploadImagesBlock = () => {
   return (
     <div className="border border-sky-light rounded-lg p-4 grid grid-cols-2 gap-4">
-      <UploadedImage />
-      <UploadedImage />
-      <UploadedImage />
-      <UploadedImage />
-      <UploadedImage />
+      <UploadedImage img={`${gcpStorageUrl}/images/rika-mei.jpg`} />
+      <UploadedImage img={`${gcpStorageUrl}/images/86Lena_shirabi_ep16.jpg`} />
```
Comment (paphonb) [inline app/shop/components/UploadImagesBlock.tsx:46]: This is not final, right?

File: app/api/uploadURL.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    projectId: gcpStorageProjectId,
+    credentials: {
+      client_email: gcpStorageClientEmail,
+      private_key: gcpStoragePrivateKey,
+    },
+  })
+
+  const bucket = gc.bucket(gcpStorageBucketName)
+  const filename = req.query.file as string
+  const folder = req.query.dir as string
```
Comment (paphonb) [inline app/api/uploadURL.ts:22]: File name and folder name shouldn't be a user input, else users can override other users' files. 

I propose a name format:
- folder name: /users/[userId]
- file name: randomly generated uuid
And have the server generate the file name and upload url when the user wants to upload a file.

**Review by natTP — COMMENTED** [review]
- The back button here should go back to menu (now it goes back to upload image and I can replace the image after submit)
image

- When the image is reselected the old image should be deleted from cloud, but... I haven't done that yet 
- Also if an image has the same filename it'll be replaced which I also haven't handled yet...

**Comment by Anon-136** [comment]
> * The back button here should go back to menu (now it goes back to upload image and I can replace the image after submit)
>   image
> * When the image is reselected the old image should be deleted from cloud, but... I haven't done that yet
> * Also if an image has the same filename it'll be replaced which I also haven't handled yet...

Done

File: app/core/mutations/deleteUploadedImage.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+})
+
+const deleteUploadedImage = resolver.pipe(
+  resolver.zod(DeleteUploadedImage),
+  resolver.authorize(),
+  async ({ url }) => {
+    const bucket = gc.bucket(gcpStorageBucketName)
+    const file = bucket.file(url.slice(gcpStorageUrl.length + 1))
+
+    await file.delete()
```
Comment (paphonb) [inline app/core/mutations/deleteUploadedImage.ts:21]: Should be a way to enforce that the user can only delete their own files, if it's not too late

Comment (Anon-136) [reply app/core/mutations/deleteUploadedImage.ts:21]: See #43

Comment (natTP) [reply app/core/mutations/deleteUploadedImage.ts:21]: Not done yet in #43 though

**Review by paphonb — APPROVED** [review]

**Pull request merged**